pio/uart_pio_dma/uart_pio_dma.c

```diff
@@ -0,0 +1,316 @@
+    read_size = sizeof(buffer_tx) - 1;
+#if USE_PIO_FOR_RX
+    // read from pio fifo
+    channel_config_set_dreq(&config_rx, pio_get_dreq(pio_hw_rx, pio_sm_rx, false));
+    // 8-bit read from the uppermost byte of the FIFO, as data is left-justified so need to add 3. Don't forget the cast!
+    dma_channel_configure(dma_channel_rx, &config_rx, buffer_rx, (io_rw_8*)&pio_hw_rx->rxf[pio_sm_rx] + 3, read_size, true); // dma started
+#else
+    // read from uart hardware
+    channel_config_set_dreq(&config_rx, uart_get_dreq(HARD_UART_INST, false));
+    dma_channel_configure(dma_channel_rx, &config_rx, buffer_rx, &uart_get_hw(HARD_UART_INST)->dr, read_size, true); // dma started
+#endif
+#endif
+
+// setup dma for write
+#if USE_DMA_FOR_TX
+    dma_channel_tx = dma_claim_unused_channel(false);
+    if (dma_channel_tx < 0) {
+        panic("No free dma channels");
+    }
+    dma_channel_config config_tx = dma_channel_get_default_config(dma_channel_tx);
+    channel_config_set_transfer_data_size(&config_tx, DMA_SIZE_8);
+    channel_config_set_read_increment(&config_tx, true);
+    channel_config_set_write_increment(&config_tx, false);
+#if USE_PIO_FOR_RX
+    // write to pio fofo
+    channel_config_set_dreq(&config_tx, pio_get_dreq(pio_hw_tx, pio_sm_tx, true));
+    dma_channel_configure(dma_channel_tx, &config_tx, &pio_hw_rx->txf[pio_sm_tx], buffer_tx, sizeof(buffer_tx) - 1, true); // dma started
+#else
+    // write to uart hardware
+    channel_config_set_dreq(&config_tx, uart_get_dreq(HARD_UART_INST, true));
+    dma_channel_configure(dma_channel_tx, &config_tx, &uart_get_hw(HARD_UART_INST)->dr, buffer_tx, sizeof(buffer_tx) - 1, true); // dma started
+#endif
+#endif
+
+// setup pio interrupt
+#if USE_PIO_FOR_RX
+    if (irq_get_exclusive_handler(pio_irqn(pio_hw_rx, PIO_IRQ_NUM))) {
+        panic("PIO IRQ in use");
+    }
+#if USE_DMA_FOR_RX
+    irq_add_shared_handler(pio_irqn(pio_hw_rx, PIO_IRQ_NUM), pio_irq_handler, PIO_IRQ_PRIORITY);
+    pio_set_irqn_source_enabled(pio_hw_rx, PIO_IRQ_NUM, pis_sm0_rx_fifo_not_empty + pio_sm_rx, true);
+    irq_set_enabled(pio_irqn(pio_hw_rx, PIO_IRQ_NUM), true);
+#endif
+#endif
+
+// setup dma interrupt
+#if USE_DMA_FOR_RX || USE_DMA_FOR_TX
+    if (irq_get_exclusive_handler(DMA_IRQN(DMA_IRQ_NUM))) {
+        panic("DMA IRQ in use");
+    }
+    irq_add_shared_handler(DMA_IRQN(DMA_IRQ_NUM), dma_irq_handler, DMA_IRQ_PRIORITY);
+#if USE_DMA_FOR_TX
+    dma_irqn_set_channel_enabled(DMA_IRQ_NUM, dma_channel_tx, true);
+#endif
+#if USE_DMA_FOR_RX
+    dma_irqn_set_channel_enabled(DMA_IRQ_NUM, dma_channel_rx, true);
+#endif
+    irq_set_enabled(DMA_IRQN(DMA_IRQ_NUM), true);
+#endif
+
+// send data
+#if USE_DMA_FOR_TX
+    dma_channel_wait_for_finish_blocking(dma_channel_tx); // wait for tx
+#elif USE_PIO_FOR_TX
+    // write to the pio fifo
+    int count_pio_tx = 0;
+    while(count_pio_tx < sizeof(buffer_tx) - 1) {
+        uart_tx_program_putc(pio_hw_tx, pio_sm_tx, buffer_tx[count_pio_tx++]);
```

Comment:
This and the following uart_puts will overflow the fifo if dma is not enabled, need to read and write one character at a time?